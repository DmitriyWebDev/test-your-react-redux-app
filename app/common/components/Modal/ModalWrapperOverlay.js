import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

class ModalWrapperOverlay extends React.Component {
  constructor(props) {
    super(props);
    this.handleOverlayClick = this.handleOverlayClick.bind(this);
    this.modalContentRef = React.createRef();
  }

  componentDidMount() {
    document.body.style.overflow = 'hidden';
  }

  componentWillUnmount() {
    document.body.style.overflow = '';
  }

  handleOverlayClick(e) {
    const { shouldCloseOnOverlayClick, onRequestClose } = this.props;

    if (!shouldCloseOnOverlayClick) return null;

    const { target } = e;
    const modalContent = this.modalContentRef.current;
    if (target !== modalContent && target.contains(modalContent)) {
      onRequestClose();
    }
  }

  render() {
    const { modalId, modalClassName, overlayClassName, children } = this.props;

    return (
      <div className={cn('app-modal__window-overlay-wrapper', overlayClassName)} onClick={this.handleOverlayClick}>
        <div id={modalId} className={cn('app-modal__window', modalClassName)} ref={this.modalContentRef}>
          {children}
        </div>
      </div>
    );
  }
}

export default ModalWrapperOverlay;

ModalWrapperOverlay.propTypes = {
  modalId: PropTypes.string,
  modalClassName: PropTypes.string,
  overlayClassName: PropTypes.string,
  onRequestClose: PropTypes.func,
  shouldCloseOnOverlayClick: PropTypes.bool,
  children: PropTypes.node,
};

ModalWrapperOverlay.defaultProps = {
  modalId: '',
  modalClassName: '',
  overlayClassName: '',
  onRequestClose: () => {},
  shouldCloseOnOverlayClick: true,
  children: <span>Modal window default content</span>,
};
