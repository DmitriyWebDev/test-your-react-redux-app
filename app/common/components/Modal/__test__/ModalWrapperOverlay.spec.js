import React from 'react';
import { mount } from 'enzyme';
import ModalWrapperOverlay from '../ModalWrapperOverlay';

const modalWindowClassName = 'app-modal__window';
const modalWindowOverlayClassName = 'app-modal__window-overlay-wrapper';

describe('ModalWrapperOverlay', () => {
  it('renders correctly, matches the snapshot', () => {
    const output = mount(<ModalWrapperOverlay />);
    expect(output.html()).toMatchSnapshot();
  });

  /** Tests on logic */

  // Test props (test render with different props values)

  it('renders correct "modalId"', () => {
    const wrapper = mount(<ModalWrapperOverlay modalId={'id-1'} />);
    expect(wrapper.exists(`.${modalWindowClassName}#id-1`)).toBe(true);
  });

  it('renders correct "modalClassName"', () => {
    const wrapper = mount(<ModalWrapperOverlay modalClassName={'class-1'} />);
    expect(wrapper.exists(`.${modalWindowClassName}.class-1`)).toBe(true);
  });

  it('renders correct "overlayClassName"', () => {
    const wrapper = mount(<ModalWrapperOverlay overlayClassName={'overlay-class-1'} />);
    expect(wrapper.exists(`.${modalWindowOverlayClassName}.overlay-class-1`)).toBe(true);
  });

  it('renders correct child component', () => {
    const wrapper = mount(
      <ModalWrapperOverlay>
        <div id={'content'}>Some content</div>
      </ModalWrapperOverlay>,
    );
    expect(wrapper.exists('#content')).toEqual(true);
  });

  // Test events handling

  it('check the onRequestClose call', () => {
    const onRequestClose = jest.fn();
    const props = {
      onRequestClose,
    };
    const AppModalWrapperOverlayComponent = mount(<ModalWrapperOverlay {...props} />);

    AppModalWrapperOverlayComponent.simulate('click');
    expect(onRequestClose).toBeCalled();
  });

  it('check the onRequestClose call, if "shouldCloseOnOverlayClick" = false', () => {
    const onRequestClose = jest.fn();
    const props = {
      onRequestClose,
      shouldCloseOnOverlayClick: false,
    };
    const AppModalWrapperOverlayComponent = mount(<ModalWrapperOverlay {...props} />);

    AppModalWrapperOverlayComponent.simulate('click');
    expect(onRequestClose).not.toBeCalled();
  });

  // Test side effects

  it('disable page scroll when mount', () => {
    mount(<ModalWrapperOverlay />);
    expect(document.body.style.overflow === 'hidden').toBe(true);
  });

  it('enable page scroll when unmount', () => {
    const AppModalWrapperOverlayComponent = mount(<ModalWrapperOverlay />);
    AppModalWrapperOverlayComponent.unmount();
    expect(document.body.style.overflow === '').toBe(true);
  });
});
