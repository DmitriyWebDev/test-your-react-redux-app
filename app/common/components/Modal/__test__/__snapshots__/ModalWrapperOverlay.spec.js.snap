// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWrapperOverlay renders correctly, matches the snapshot 1`] = `
<div
  className="app-modal__window-overlay-wrapper overlay-class-1"
  onClick={[Function]}
>
  <div
    className="app-modal__window class-1"
    id="id-1"
  >
    <span>
      Modal window default content
    </span>
  </div>
</div>
`;
